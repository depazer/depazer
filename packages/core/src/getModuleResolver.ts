import { resolve } from 'node:path'

import { commonAdaptor } from './adaptors/commonAdaptor'
import { environmentScanner } from './environmentScanner'
import { readPackageJSON } from '@depazer/shared'

import type { ModuleObject } from '@/types/moduleGraph'

/**
 *
 * @example
 * const resolver = await getModuleResolver(resolve(), true)
 *
 * if (typeof resolver !== 'string') {
 *   const res = await resolver(4)
 *
 *   writeFile(resolve('/.json'), JSON.stringify(res, null, 2))
 *   writeFile(resolve('/flat.json'), JSON.stringify(flatPipe(res), null, 2))
 * }
 */
export async function getModuleResolver(root: string, includeDeps: boolean) {
  const { packageManager, monorepo: _, error } = await environmentScanner(root)
  if (typeof error === 'string') return error

  const moduleObject = await initModuleObject(root, includeDeps)

  return (depth: number = Infinity) => commonAdaptor(packageManager, root, moduleObject, depth)
}

export async function initModuleObject(root: string, includeDeps: boolean) {
  const rootPackageJson = resolve(root, 'package.json')

  const { name, version, dependencies, devDependencies } = await readPackageJSON(rootPackageJson)
  return formatModuleObject(name, version, dependencies, includeDeps ? devDependencies : undefined)
}

function formatModuleObject(
  name: string,
  version: string,
  dependencies: Record<string, string>,
  devDependencies?: Record<string, string>
): ModuleObject {
  return {
    name,
    version,
    dependencies: Object.entries(dependencies).map(([name, version]) => ({
      name,
      version,
      dependencies: []
    })),
    devDependencies: Object.entries(devDependencies ?? {}).map(([name, version]) => ({
      name,
      version,
      dependencies: []
    }))
  }
}
