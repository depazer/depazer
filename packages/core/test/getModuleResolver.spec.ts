import { getModuleResolver } from '@/getModuleResolver'
import { describe, test } from 'vitest'

describe('getModuleResolver', () => {
  test.todo('should return an environment object', async ({ expect }) => {})
})
