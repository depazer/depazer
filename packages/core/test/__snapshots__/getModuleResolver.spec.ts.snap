// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`getModuleResolver > should return a module object without devDependency 1`] = `
{
  "dependencies": [
    {
      "dependencies": [],
      "name": "test-dep",
      "version": "unknown",
    },
  ],
  "devDependencies": [],
  "name": "test",
  "version": "1.0.0",
}
`;

exports[`getModuleResolver > should return module Object with devDeps 1`] = `
{
  "dependencies": [
    {
      "dependencies": [],
      "name": "test-dep",
      "version": "unknown",
    },
  ],
  "devDependencies": [
    {
      "dependencies": [],
      "name": "test-dev-dep",
      "version": "unknown",
    },
  ],
  "name": "test",
  "version": "1.0.0",
}
`;
